metadetect/shearpos.py

```diff
@@ -0,0 +1,144 @@
+import numpy as np
+import galsim
+
+
+def shear_positions(rows, cols, shear_str, obs, step=0.01):
+    """
+    unshear the input row and column positions
+
+    Parameters
+    ----------
+    rows: array
+        array of row values in sheared coordinates
+    cols: array
+        array of col values in sheared coordinates
+    shear_str: string
+        'noshear', '1p', '1m', '2p', '2m'
+    jac: ngmix.Jacobian
+        Describes the wcs
+    """
+
+    if shear_str == 'noshear':
+        return rows, cols
+
+    shear = get_galsim_shear(shear_str, step)
+
+    # this is the matrix that does shearing in u, v
+    a = shear.getMatrix()
+
+    # this is the jacobian that deals with the WCS
+    jac = obs.jacobian
+
+    # we need the canonical image center in (u, v) for undoing the
+    # shearing
+
+    dims = obs.image.shape
+    row_cen = (dims[0] - 1) / 2
+    col_cen = (dims[1] - 1) / 2
+
+    v_cen, u_cen = jac.get_vu(row=row_cen, col=col_cen)
+
+    def _shearpos(row, col):
+        # apply WCS to get to world coords
+        v, u = jac.get_vu(row=row, col=col)
+
+        # unshear (subtract and then add the canonical center in u, v)
+        u = np.atleast_1d(u) - u_cen
+        v = np.atleast_1d(v) - v_cen
+        out = np.dot(a, np.vstack([u, v]))
+        assert out.shape[1] == u.shape[0]
+        u_sheared = out[0] + u_cen
+        v_sheared = out[1] + v_cen
+
+        # undo the WCS to get to image coords
+        row_sheared, col_sheared = jac.get_rowcol(v=v_sheared, u=u_sheared)
+        return row_sheared[0], col_sheared[0]
+
+    rows_sheared = np.zeros(rows.size)
+    cols_sheared = np.zeros(rows.size)
+
+    for i in range(rows.size):
+        rows_sheared[i], cols_sheared[i] = _shearpos(rows[i], cols[i])
```

Comment:
i forget why I coded it this way. This is not a bottleneck, but vectorizing this loop would be nice in an ideal world.

Reply:
the way the code works now, it crashes if you send arrays.  Let me know if you know how to modify it

Reply:
np.dot should do the trick as long as the axes are ordered properly.

Reply:
I think the right order would be `np.dot(pos, shear_mat)` where pos is `(n_obj, 2)` and shear mat is `(2, 2)` this would do the product-sum over the last axis of pos and the first axis of th shear mat. The best way to find out is to test OFC. 

TBH, let's just leave this for now. We can do it later.

Reply:
Nope the above is wrong. You want no.dot(shearmat, pos.T).T for pos as above I think

Reply:
I added the vectorized version